src/psyclone/tests/alggen_test.py

```diff
@@ -429,3 +430,212 @@ def test_zero_invoke_dynamo0p1():
+
+    param str code: Fortran code in a string
+
+    :returns: parse tree of the supplied code.
+    :rtype: :py:class:`fparser.two.utils.Base`
+
+    '''
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.parser import ParserFactory
+    reader = FortranStringReader(code)
+    f2003_parser = ParserFactory().create()
+    return f2003_parser(reader)
+
+# function adduse tests These should be moved in #240.
+
+
+def test_adduse_location_none():
+    '''Test that the expected exception is raised when the specified
+    location is None. There is a check for this as the parse tree can
+    contain nodes with the value None.
+
+    '''
+    location = None
+    with pytest.raises(GenerationError) as excinfo:
+        _ = adduse(None, location, None)
+    assert "Location argument must not be None." \
+        in str(excinfo.value)
+
+
+def test_adduse_only_names1():
+    '''Test that the expected output is obtained in a Fortran program when
+    variable/function names are provided for a use statement and only
+    is True.
+
+    '''
+    parse_tree = get_parse_tree(CODE)
+    location = parse_tree.content[0].content[0]
+    name = "my_use"
+
+    new_parse_tree = adduse(parse_tree, location, name, only=True,
+                            funcnames=["a", "b", "c"])
+    assert "PROGRAM test\n  USE my_use, ONLY: a, b, c\n  INTEGER :: i\n" \
+        in str(new_parse_tree)
+
+
+def test_adduse_only_names2():
+    '''Test that the expected output is obtained in a Fortran subroutine
+    when variable/function names are provided for a use statement and
+    only is True.
+
+    '''
+    parse_tree = get_parse_tree(
+        "subroutine test()\n"
+        "  integer :: i\n"
+        "  i=0\n"
+        "end subroutine test\n")
+    location = parse_tree.content[0].content[0]
+    name = "my_use"
+
+    new_parse_tree = adduse(parse_tree, location, name, only=True,
+                            funcnames=["a", "b", "c"])
+    assert ("SUBROUTINE test\n  USE my_use, ONLY: a, b, c\n"
+            "  INTEGER :: i\n") in str(new_parse_tree)
+
+
+def test_adduse_only_nonames():
+    '''Test that the expected output is obtained when no variable/function
+    names are provided for a use statement and only is True.
+
+    '''
+    parse_tree = get_parse_tree(CODE)
+    location = parse_tree.content[0].content[0]
+    name = "my_use"
+
+    new_parse_tree = adduse(parse_tree, location, name, only=True)
+    assert "PROGRAM test\n  USE my_use, ONLY:\n  INTEGER :: i\n" \
+        in str(new_parse_tree)
+
+
+def test_adduse_noonly_names():
+    '''Test that the expected output is obtained when variable/function
+    names are provided for a use statement and only is False.
+
+    '''
+    parse_tree = get_parse_tree(CODE)
+    location = parse_tree.content[0].content[0]
+    name = "my_use"
+    with pytest.raises(GenerationError) as excinfo:
+        _ = adduse(parse_tree, location, name, funcnames=["a", "b", "c"])
```

Comment:
OK, I've changed it so that adding funcnames with content implies only if it is not explicitely added, however we raise an exception if it is explicitly set to False. If it is set to True without any funcnames content (or no funcnames argument supplied) then we get an only with no funcnames).

Reply:
Great, thanks.